design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+# Crossplane v2
+
+* Owner: Nic Cope (@negz)
+* Reviewers: Crossplane Steering Committee
+* Status: Draft
+
+
+## Executive Summary
+
+This document proposes a new version of Crossplane that’s simpler and better
+suited to managing applications, not just the infrastructure they depend on.
+
+In Crossplane v2:
+
+* Most composite resources (XRs) are namespaced
+* All managed resources (MRs) are namespaced
+* XRs can compose any Kubernetes resource - not just MRs
+* Claims are removed
+
+Under this proposal Crossplane v2 would be backward compatible with Crossplane
+v1. Most Crossplane users could upgrade to Crossplane v2 without any breaking
+changes.
+
+Existing v1 style cluster scoped XRs with claims and cluster scoped MRs would
+continue to work in Crossplane v2. These would be considered legacy features and
+wouldn’t appear in Crossplane’s main documentation flows for new users. Existing
+users could adopt v2 style XRs and MRs at their leisure.
```

Comment:
Is the idea that existing users could adopt V2 style resources without breaking changes on current V1 style resources? I also wonder if in terms of the providers if there will be a point where providers will want to kill of the V1 style implementation of MRs and only support V2 so would love to understand what the thoughts are for that?

Reply:
Yes exactly. We're adding a new way to do things but keeping the old one. This way folks can start using the new way without forcing an immediate migration.

I do think the old way (e.g. `LegacyCluster` XRs and cluster scoped MRs) will go away eventually. It feels too early to make a specific plan/timeline for that though.

In particular we're not going to have any v1 XR/MR -> v2 XR/MR migration tools at launch of v2. I'd like to have such a tool but it'll come after launch.